packages/foam-vscode/src/core/model/workspace.ts

```diff
@@ -112,36 +112,49 @@ export class FoamWorkspace implements IDisposable {
         }
       }
     }
-    const identifier = getShortestIdentifier(
+    let identifier = getShortestIdentifier(
       forResource.path,
       amongst.map(uri => uri.path)
     );
 
-    return identifier.endsWith('.md') ? identifier.slice(0, -3) : identifier;
+    identifier = identifier.endsWith('.md')
+      ? identifier.slice(0, -3)
+      : identifier;
+
+    if (forResource.fragment) {
+      identifier += `#${forResource.fragment}`;
+    }
+
+    return identifier;
   }
 
   public find(resourceId: URI | string, reference?: URI): Resource | null {
     const refType = getReferenceType(resourceId);
-    switch (refType) {
-      case 'uri':
-        const uri = resourceId as URI;
-        return this.exists(uri)
-          ? this.resources.get(normalize(uri.path)) ?? null
-          : null;
+    if (refType === 'uri') {
+      const uri = resourceId as URI;
+      return this.exists(uri)
+        ? this.resources.get(normalize(uri.path)) ?? null
+        : null;
+    }
```

Comment:
I believe you can reorganize the code in this function in a different way:

```ts
switch (refType) {
      case 'key':
           ...
           return sorted[0];
      case 'absolute-path':
           ...
           resourceUri = ...;
           break;
      case 'relative-path':
           ...
           resourceUri = ...;
           break;
      case 'uri':
           resourceUri = resourceId;
           break;
}
return this.exists(resourceUri)
        ? this.resources.get(normalize(resourceUri.path)) ?? null
        : null;
```

This:
- Makes it clearer that ultimately what is resolved in the workspace is a uri.
- Deduplicates some normalization & resolution logic.
- Deals with the confusing fact of having three names (`uri`, `resourceUri` and `targetUri`) for what is essentially the same thing.

Unfortunately the `key` case has to be dealt in a different way because `listById` already returns resources, but all in all I think this will be an improvement.

Reply:
Also the code:

```ts
      return this.exists(uri)
        ? this.resources.get(normalize(uri.path)) ?? null
        : null;
```

in the `uri` clause seems a bit redundant: `exists` checks if the uri is a placeholder and then if `this.resources.get(normalize(uri.path))` is not null. So here you can just check if the uri is a placeholder, since `this.resources.get(normalize(uri.path))` will return null anyway when not found. This way you avoid the double resolution. Moreover, I don't believe the `?? null` is adding anything.

Reply:
Indeed `exists` is only used here and doesn't add too much to the API. After the change above, you can remove `exists` since it's easily replaced by `isSome(ws.find(uri))`, or keep it in the `API` just like an unchecked version of `ws.get(uri)`:

```ts
  public exists(uri: URI): boolean {  // <-- doesn't add too much, maybe remove it
    return isSome(this.find(uri));
  }

  public get(uri: URI): Resource {
    const note = this.find(uri);
    if (isSome(note)) {
      return note;
    } else {
      throw new Error('Resource not found: ' + uri.path);
    }
  }
```

Reply:
Good pointers, I have implemented some in the PR